data/completion/etelpmoc.sh

```diff
@@ -0,0 +1,195 @@
+#!/bin/bash
+
+# etelpmoc is the backwards half of complete: it de-serialises the tab
+# completion request into the appropriate environs expected by the tab
+# completion tools, performs whatever action is wanted, and serialises the
+# result. It accomplishes this by having functions override the builtin
+# completion commands.
+#
+# this always runs "inside", in the same environment you get when doing "snap
+# run --shell", and snap-exec is the one setting the first argument to the
+# completion script set in the snap. The rest of the arguments come through
+# from snap-run --command=complete <snap> <args...>
+
+_die() {
+    echo "$*" >&2
+    exit 1
+}
+
+if [[ "$BASH_SOURCE" != "$0" ]]; then
+    _die "ERROR: this is meant to be run, not sourced."
+fi
+
+if [[ "${#@}" -lt 8 ]]; then
+    _die "USAGE: $0 <script> <COMP_TYPE> <COMP_KEY> <COMP_POINT> <COMP_CWORD> <COMP_WORDBREAKS> <COMP_LINE> cmd [args...]"
+fi
+
+# De-serialize the command line arguments and populate tab completion environment
+_compscript="$1"
+shift
+COMP_TYPE="$1"
+shift
+COMP_KEY="$1"
+shift
+COMP_POINT="$1"
+shift
+COMP_CWORD="$1"
+shift
+COMP_WORDBREAKS="$1"
+shift
+# duplication, but whitespace is eaten and that throws off COMP_POINT
+COMP_LINE="$1"
+shift
+# rest of the args is the command itself
+COMP_WORDS=("$@")
+
+COMPREPLY=()
+
+if [[ ! "$_compscript" ]]; then
+    _die "ERROR: completion script filename can't be empty"
+fi
+if [[ ! -f "$_compscript" ]]; then
+    _die "ERROR: completion script does not exist"
+fi
+
+# Source the bash-completion library functions and common completion setup
+. /usr/share/bash-completion/bash_completion
+# Now source the snap's 'completer' script itself
+. "$_compscript"
+
+# _compopts is an associative array, which keys are options. The options are
+# described in bash(1)'s description of the -o option to the "complete"
+# builtin, and they affect how the completion options are presented to the user
+# (e.g. adding a slash for directories, whether to add a space after the
+# completion, etc). These need setting in the user's environment so need
+# serializing separately from the completions themselves.
+declare -A _compopts
+
+# wrap compgen, setting _compopts for any options given.
+# (as these options need handling separately from the completions)
+compgen() {
+    local opt
+
+    while getopts :o: opt; do
+        case "$opt" in
+            o)
+                _compopts["$OPTARG"]=1
+                ;;
+        esac
+    done
+    builtin compgen "$@"
+}
+
+# compopt replaces the original compopt with one that just sets/unsets entries
+# in _compopts
+compopt() {
+    local i
+
+    for ((i=0; i<$#; i++)); do
+        case "${!i}" in
+            -o)
+                ((i++))
+                _compopts[${!i}]=1
```

Comment:
This syntax is pretty unusual. I'd leave a comment for the future readers to explain what is going on.